File: hows-ts/src/parse.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import { decodeFrame, encodeFrame, type Frame, FrameType } from './parse.js'
import { randomBigInt } from './util.js'

test('encode then decode result matches original frame', () => {
	const frame: Frame = {
		type: FrameType.REQUEST,
		requestId: 11n,
		request: {
			m: 'GET',
			u: '/api/test.json',
			h: [
				['Content-Type', 'text/plain'],
				['X-Forwarded-For', '1.1.1.1'],
			],
		},
	}
	const enc1 = encodeFrame(frame)
	const dec1 = decodeFrame(enc1)

	assert.deepStrictEqual(frame, dec1)
})

test('JSON codec benchmark', () => {
	const start = performance.now()

	for (let i = 0; i < 10_000; i++) {
		const encoded = encodeFrame({
			type: FrameType.REQUEST,
			requestId: randomBigInt(),
			request: {
				m: 'POST',
				u: '/api/service/TestThing',
				h: [
					['Content-Type', 'application/json'],
					['Content-Length', '255'],
					['User-Agent', 'Firefox'],
				],
			},
		})
		decodeFrame(encoded)
	}

	const end = performance.now()

	console.log(`took ${(end - start) / 1_000}s`)
})

test('JSON decode benchmark', () => {
	const start = performance.now()

	const encoded = encodeFrame({
		type: FrameType.REQUEST,
		requestId: randomBigInt(),
		request: {
			m: 'POST',
			u: '/api/service/TestThing',
			h: [
				['Content-Type', 'application/json'],
				['Content-Length', '255'],
				['User-Agent', 'Firefox'],
			],
		},
	})

	for (let i = 0; i < 10_000; i++) {
		decodeFrame(encoded)
	}

	const end = performance.now()

	console.log(`took ${(end - start) / 1_000}s`)
})

test('JSON encode benchmark', () => {
	const start = performance.now()

	for (let i = 0; i < 10_000; i++) {
		encodeFrame({
			type: FrameType.REQUEST,
			requestId: randomBigInt(),
			request: {
				m: 'POST',
				u: '/api/service/TestThing',
				h: [
					['Content-Type', 'application/json'],
					['Content-Length', '255'],
					['User-Agent', 'Firefox'],
				],
			},
		})
	}

	const end = performance.now()

	console.log(`took ${(end - start) / 1_000}s`)
})

File: hows-go/parse_test.go
package hows

import (
	"math/rand/v2"
	"testing"
)

func TestEncodeThenDecode(t *testing.T) {
	f := Frame{
		Type:      FrameTypeRequest,
		RequestId: 11,
		Request: FrameRequest{
			Method: "GET",
			Uri:    "/api/test.json",
			Headers: [][]string{
				{"Content-Type", "text/plain"},
				{"X-Forwarded-For", "1.1.1.1"},
			},
		},
	}

	enc, err := EncodeFrame(f)
	if err != nil {
		t.Fatal(err)
	}

	dec, err := DecodeFrame(enc)
	if err != nil {
		t.Fatal(err)
	}

	if dec.Type != f.Type {
		t.Fatalf("wrong frame type: %c", dec.Type)
	}
	if dec.RequestId != f.RequestId {
		t.Fatalf("wrong request ID: %d", dec.RequestId)
	}
	req := f.Request
	decReq := dec.Request
	if decReq.Method != req.Method {
		t.Fatalf("wrong method: %s", decReq.Method)
	}
	if decReq.Uri != req.Uri {
		t.Fatalf("wrong URI: %s", decReq.Uri)
	}
	if len(decReq.Headers) != len(req.Headers) {
		t.Fatalf("wrong header count: %d", len(decReq.Headers))
	}
}

func BenchmarkEncodeThenDecode(b *testing.B) {
	for b.Loop() {
		enc, err := EncodeFrame(Frame{
			Type:      FrameTypeRequest,
			RequestId: rand.Int64(),
			Request: FrameRequest{
				Method: "POST",
				Uri:    "/api/service/TestThing",
				Headers: [][]string{
					{"Content-Type", "application/json"},
					{"Content-Length", "255"},
					{"User-Agent", "Firefox"},
				},
			},
		})
		if err != nil {
			b.Fatal(err)
		}

		_, _ = DecodeFrame(enc)
	}
}
